File: benchmarks/go/binary_trees.go
// The Computer Language Benchmarks Game
// https://benchmarksgame-team.pages.debian.net/benchmarksgame/
//
// Binary trees: build and checksum complete binary trees

package main

import (
	"fmt"
	"os"
	"strconv"
)

type Node struct {
	value int64
	left  *Node
	right *Node
}

func bottomUpTreeWithValue(depth int, value int64) *Node {
	if depth <= 0 {
		return &Node{value: value}
	}
	next := value * 2
	return &Node{
		left:  bottomUpTreeWithValue(depth-1, next),
		right: bottomUpTreeWithValue(depth-1, next+1),
	}
}

func bottomUpTree(depth int) *Node {
	return bottomUpTreeWithValue(depth, 1)
}

func itemCheck(node *Node) int64 {
	if node.left == nil {
		return node.value
	}
	return 1 + itemCheck(node.left) + itemCheck(node.right)
}

func main() {
	n := 15
	if len(os.Args) > 1 {
		if v, err := strconv.Atoi(os.Args[1]); err == nil {
			n = v
		}
	}

	minDepth := 4
	maxDepth := n
	if minDepth+2 > n {
		maxDepth = minDepth + 2
	}

	stretchDepth := maxDepth + 1
	stretchTree := bottomUpTree(stretchDepth)
	fmt.Printf("stretch tree of depth %d\t check: %d\n", stretchDepth, itemCheck(stretchTree))

	longLivedTree := bottomUpTree(maxDepth)

	for depth := minDepth; depth <= maxDepth; depth += 2 {
		iterations := 1 << uint(maxDepth-depth+minDepth)
		check := int64(0)
		for i := 1; i <= iterations; i++ {
			t := bottomUpTree(depth)
			check += itemCheck(t)
		}
		fmt.Printf("%d\t trees of depth %d\t check: %d\n", iterations, depth, check)
	}

	fmt.Printf("long lived tree of depth %d\t check: %d\n", maxDepth, itemCheck(longLivedTree))
}

File: benchmarks/c/binary_trees.c
/* The Computer Language Benchmarks Game
   https://benchmarksgame-team.pages.debian.net/benchmarksgame/
   contributed by Kevin Carson */
#include <stdio.h>
#include <stdlib.h>

typedef struct node {
    struct node *left, *right;
    long long value;
} Node;

static Node *make_with_value(int depth, long long value) {
    Node *n = malloc(sizeof(Node));
    if (depth == 0) {
        n->left = n->right = NULL;
        n->value = value;
    } else {
        long long next = value * 2;
        n->left = make_with_value(depth - 1, next);
        n->right = make_with_value(depth - 1, next + 1);
        n->value = 0;
    }
    return n;
}

static Node *make(int depth) {
    return make_with_value(depth, 1);
}

static long long check(Node *n) {
    if (!n->left) return n->value;
    return 1 + check(n->left) + check(n->right);
}

static void free_tree(Node *n) {
    if (n->left) { free_tree(n->left); free_tree(n->right); }
    free(n);
}

int main(int argc, char **argv) {
    int min_depth = 4;
    int n = argc > 1 ? atoi(argv[1]) : 10;
    int max_depth = n;
    if (min_depth + 2 > max_depth) max_depth = min_depth + 2;
    int stretch_depth = max_depth + 1;

    Node *stretch = make(stretch_depth);
    printf("stretch tree of depth %d\t check: %lld\n", stretch_depth, check(stretch));
    free_tree(stretch);

    Node *long_lived = make(max_depth);

    for (int d = min_depth; d <= max_depth; d += 2) {
        int iterations = 1 << (max_depth - d + min_depth);
        long long total = 0;
        for (int i = 0; i < iterations; i++) {
            Node *t = make(d);
            total += check(t);
            free_tree(t);
        }
        printf("%d\t trees of depth %d\t check: %lld\n", iterations, d, total);
    }

    printf("long lived tree of depth %d\t check: %lld\n", max_depth, check(long_lived));
    free_tree(long_lived);
    return 0;
}
